zerver/lib/actions.py

```diff
@@ -817,6 +817,67 @@ def do_send_messages(messages):
+    # type: (Dict[str, Any]) -> None
+    if notification['recipient'].type == Recipient.PERSONAL:
+        notification['recipients'] = [notification['sender'],
+                                      get_user_profile_by_id(notification['recipient'].type_id)]
+    elif notification['recipient'].type == Recipient.HUDDLE:
+        # We use select_related()/only() here, while the PERSONAL case above uses
+        # get_user_profile_by_id() to get UserProfile objects from cache.  Streams will
+        # typically have more recipients than PMs, so get_user_profile_by_id() would be
+        # a bit more expensive here, given that we need to hit the DB anyway and only
+        # care about the email from the user profile.
+        fields = [
+            'user_profile__id',
+            'user_profile__email',
+            'user_profile__is_active',
+            'user_profile__realm__domain'
+        ]
+        query = Subscription.objects.select_related("user_profile", "user_profile__realm").only(*fields).filter(
+                recipient=notification['recipient'], active=True)
+        notification['recipients'] = [s.user_profile for s in query]
+    else:
+        raise ValueError('Bad recipient type')
+
+    # Only deliver the message to active user recipients
+    notification['active_recipients'] = [user_profile for user_profile in notification['recipients']
+                                         if user_profile.is_active]
+
+    event = dict(
+            type            = 'typing',
+            op              = notification['op'],
+            sender          = notification['sender'],
+            recipients      = notification['recipients'])
+
+    send_event(event, notification['active_recipients'])
+
+# check_send_typing_notification:
+# Checks the typing notification and sends it
+def check_send_typing_notification(sender, notification_to, operator):
+    # type: (UserProfile, Sequence[text_type], text_type) -> None
+    typing_notification = check_typing_notification(sender, notification_to, operator)
+    do_send_typing_notification(typing_notification)
+
+# check_typing_notification:
+# Returns typing notification ready for sending with do_send_typing_notification on success
+# or the error message (string) on error.
+def check_typing_notification(sender, notification_to, operator):
+    # type: (UserProfile, Sequence[text_type], text_type) -> Dict[str, Any]
+    if len(notification_to) == 0:
+        raise JsonableError(_('Missing parameter: \'to\' (recipient)'))
+    elif operator not in ('start', 'stop'):
+        raise JsonableError(_('Invalid \'op\' value (should be start or stop)'))
+    else:
+        try:
+            recipient = recipient_for_emails(notification_to, False,
+                                             sender, sender)
+        except ValidationError as e:
+            assert isinstance(e.messages[0], six.string_types)
+            raise JsonableError(e.messages[0])
```

Comment:
Can you add a negative test for this condition (trying to send a typing notification to a user who doesn't exist)?